tools/topology/sof-byt-codec.m4

```diff
@@ -94,9 +94,12 @@ DAI_CONFIG(SSP, SSP_NUM, 0, SSP2-Codec,
 		      SSP_TDM(2, 25, 3, 3),
 		      SSP_CONFIG_DATA(SSP, SSP_NUM, 24)))
 
-VIRTUAL_WIDGET(ssp2 Rx, out_drv, 1)
-VIRTUAL_WIDGET(ssp2 Tx, out_drv, 2)
-VIRTUAL_WIDGET(ssp0 Tx, out_drv, 3)
+ifelse(SSP_NUM, 2,
+`VIRTUAL_WIDGET(ssp2 Rx, out_drv, 1)
+VIRTUAL_WIDGET(ssp2 Tx, out_drv, 2)', ,)
+
+ifelse(SSP_NUM, 0,
+`VIRTUAL_WIDGET(ssp0 Tx, out_drv, 3)
 VIRTUAL_WIDGET(ssp0 Rx, out_drv, 4)
 VIRTUAL_DAPM_ROUTE_IN(modem_in, SSP, 0, IN, 0)
-VIRTUAL_DAPM_ROUTE_OUT(modem_out, SSP, 0, OUT, 1)
+VIRTUAL_DAPM_ROUTE_OUT(modem_out, SSP, 0, OUT, 1)', ,)
```

Comment:
@plbossart @juimonen it was probably a bit of an oversight when merging the changes for byt-cr. But do we need these virtual dapm routes at all? The virtual widgets are good enough to handle the machine driver warnings

Reply:
@ranj063 yes those routes are required. See dmesg excerpt below

````
[  115.346225] bytcr_rt5640 bytcr_rt5640: ASoC: no source widget found for modem_out
[  115.346234] bytcr_rt5640 bytcr_rt5640: ASoC: Failed to add route modem_out -> direct -> ssp0 Tx
[  115.346243] bytcr_rt5640 bytcr_rt5640: ASoC: no sink widget found for modem_in
[  115.346248] bytcr_rt5640 bytcr_rt5640: ASoC: Failed to add route ssp0 Rx -> direct -> modem_in
[  115.346267] bytcr_rt5640 bytcr_rt5640: ASoC: failed to init SSP2-Codec: -19
[  115.346305
````

Reply:
@plbossart I think these errors are because you dont have the vritual widgets for "modem_out" and modem_in

Reply:
that's possible. let me check.

Reply:
you're right @ranj063 

all I need is this

````
VIRTUAL_WIDGET(ssp2 Rx, out_drv, 1)
VIRTUAL_WIDGET(ssp2 Tx, out_drv, 2)
VIRTUAL_WIDGET(ssp0 Tx, out_drv, 3)
VIRTUAL_WIDGET(ssp0 Rx, out_drv, 4)
VIRTUAL_WIDGET(modem_out, out_drv, 5)
VIRTUAL_WIDGET(modem_in, out_drv, 6)
````

I wonder if we should just add them instead to byt.m4 and cht.m4

Also not sure why it's all out_drv even for capture parts?

Reply:
@plbossart we could add them to byt/cht.m4 but these arent needed for the nocodec topology. But they are harmless.

Also, the type doesnt really matter as long its one that the SOF driver ignores.

Reply:
let's move them to common parts then, it's easier to have all of those in one place.